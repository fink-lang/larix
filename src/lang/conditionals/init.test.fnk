{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'

{parse} = import '../../parser.fnk'


describe 'match ...:', fn:

  it 'parses simple', fn:
    expect
      parse '
        match item:
          test_pattern: result

          # fallback if no match
          else: spam
          # trailing comment
        foobar
      '
      to_match_snapshot


  it 'parses complex: {foo}: bar', fn:
    expect
      parse '
        match item:
          # nested record pattern
          {foo, spam: {ni}}: bar

          # nested iterables
          [foo, [1, 2]]: bar

          # mixed nested:
          [foo, {bar: 123, spam: [456, 789]}]: ni

          else:
            # fallback for no match
            spam
      '
      to_match_snapshot


  it 'parses partials', fn:
    expect
      parse '
        match foo:
          {bar: ? > 1}: ni

          is_str ?: foo

          [..., ?] > shrub: ni
      '
      to_match_snapshot
