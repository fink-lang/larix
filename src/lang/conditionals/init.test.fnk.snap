// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match ...: parses complex: {foo}: bar 1`] = `
module (1:0-14:0)
  block match (1:0-14:0)
    ident (1:6-1:10) item
    :
    match:expr (3:2-3:24)
      # nested record pattern
      rec (3:2-3:19)
        rec:kv (3:3-3:6)
          ident (3:3-3:6) foo
        rec:kv (3:8-3:18)
          ident (3:8-3:12) spam
          rec (3:14-3:18)
            rec:kv (3:15-3:17)
              ident (3:15-3:17) ni
      block (3:19-3:24)
        ident (3:21-3:24) bar
    match:expr (6:2-6:20)
      # nested iterables
      list (6:2-6:15)
        ident (6:3-6:6) foo
        list (6:8-6:14)
          number (6:9-6:10) 1
          number (6:12-6:13) 2
      block (6:15-6:20)
        ident (6:17-6:20) bar
    match:expr (9:2-9:41)
      # mixed nested:
      list (9:2-9:37)
        ident (9:3-9:6) foo
        rec (9:8-9:36)
          rec:kv (9:9-9:17)
            ident (9:9-9:12) bar
            number (9:14-9:17) 123
          rec:kv (9:19-9:35)
            ident (9:19-9:23) spam
            list (9:25-9:35)
              number (9:26-9:29) 456
              number (9:31-9:34) 789
      block (9:37-9:41)
        ident (9:39-9:41) ni
    block else (11:2-14:0):
      # fallback for no match
      ident (13:4-13:8) spam
`;

exports[`match ...: parses partials 1`] = `
module (1:0-7:0)
  block match (1:0-7:0)
    ident (1:6-1:9) foo
    :
    match:expr (2:2-2:18)
      rec (2:2-2:14)
        rec:kv (2:3-2:13)
          ident (2:3-2:6) bar
          comp > (2:8-2:13)
            partial (2:8-2:9) ?
            number (2:12-2:13) 1
      block (2:14-2:18)
        ident (2:16-2:18) ni
    match:expr (4:2-4:15)
      call (4:2-4:10)
        ident (4:2-4:8) is_str
        partial (4:9-4:10) ?
      block (4:10-4:15)
        ident (4:12-4:15) foo
    match:expr (6:2-7:0)
      comp > (6:2-6:18)
        list (6:2-6:10)
          spread ... (6:3-6:6)
          partial (6:8-6:9) ?
        ident (6:13-6:18) shrub
      block (6:18-7:0)
        ident (6:20-6:22) ni
`;

exports[`match ...: parses simple 1`] = `
module (1:0-8:0)
  block match (1:0-5:12)
    ident (1:6-1:10) item
    :
    match:expr (2:2-2:22)
      ident (2:2-2:14) test_pattern
      block (2:14-2:22)
        ident (2:16-2:22) result
    # fallback if no match
    block else (5:2-5:12):
      ident (5:8-5:12) spam
  ident (7:0-7:6) foobar
`;
